src/k2/module/k23si/Module.cpp

```diff
@@ -272,6 +277,285 @@ bool K23SIPartitionModule::_validateStaleWrite(dto::K23SIWriteRequest& request,
+                // Request's payload skipped this value(means the field is updated to NULL), AND
+                // base payload has this field.
+                // Then exclude this field, at the mean time _advancePosition of base payload.
+                request.value.excludedFields[i] = true;
+                if (!_advancePayloadPosition(basePayload, schema.fields[i].type)) return false;
+            } else {
+                // Request's payload skipped this value, AND base payload also skipped this field.
+                // set excludedFields[i]
+                request.value.excludedFields[i] = true;
+            }
+        } else {
+            // this field is NOT updated
+            if (request.value.excludedFields[i] == 0 &&
+                    (versions[0].value.excludedFields.empty() || versions[0].value.excludedFields[i] == 0)) {
+                // Request's payload contains this field, AND
+                // base SKVRecord also has value of this field.
+                // copy 'base skvRecord' value, at the mean time _advancePosition of 'req' payload.
+                if (!_copyPayloadBaseToUpdate(basePayload, payload, schema.fields[i].type)) return false;
+                if (!_advancePayloadPosition(request.value.fieldData, schema.fields[i].type)) return false;
+            } else if (request.value.excludedFields[i] == 0 && 
+                    (!versions[0].value.excludedFields.empty() && versions[0].value.excludedFields[i] == 1)) {
+                // Request's payload contains this field, AND
+                // base SKVRecord do NOT has this field.
+                // skip this field, at the mean time _advancePosition of 'req' payload.
+                request.value.excludedFields[i] = true;
+                if (!_advancePayloadPosition(request.value.fieldData, schema.fields[i].type)) return false;
+            } else if (request.value.excludedFields[i] == 1 && 
+                    (versions[0].value.excludedFields.empty() || versions[0].value.excludedFields[i] == 0)) {
+                // Request's payload do NOT contain this field, AND
+                // base SKVRecord has value of this field.
+                // copy 'base skvRecord' value.
+                if (!_copyPayloadBaseToUpdate(basePayload, payload, schema.fields[i].type)) return false;
+                request.value.excludedFields[i] = false;
+            } else {
+                // else, request payload skipped this field, AND base SKVRecord also skipped this field,
+                // set excludedFields[i]
+                request.value.excludedFields[i] = true;
+            }
+        }        
+    }
+    
+    request.value.fieldData = std::move(payload);
+    request.value.fieldData.truncateToCurrent();
+    return true;
+}
+
+bool K23SIPartitionModule::_makeFieldsForDiffVersion(dto::Schema schema, dto::Schema baseSchema, dto::K23SIPartialUpdateRequest& request, std::deque<dto::DataRecord>& versions) {
```

Comment:
You can pass the schemas by reference to avoid a memory copy, and it might be cleaner to pass a reference to the first record version instead of passing the whole deque here. Same comment for _makeFieldsForSameVersion above